pkg/cgroups/cgroupsv2/runtime.go

```diff
@@ -0,0 +1,147 @@
+	jobv1 "github.com/kralicky/jobserver/pkg/apis/job/v1"
+	"github.com/kralicky/jobserver/pkg/cgroups"
+	"github.com/kralicky/jobserver/pkg/jobs"
+	"github.com/kralicky/jobserver/pkg/util"
+)
+
+const gracePeriod = 10 * time.Second
+
+type v2Runtime struct {
+	mgr *cgroupManager
+}
+
+func NewRuntime() (jobs.Runtime, error) {
+	mgr, err := newCgroupManager()
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup jobserver cgroup: %w", err)
+	}
+
+	return &v2Runtime{
+		mgr: mgr,
+	}, nil
+}
+
+// Execute implements jobs.Runtime.
+func (l *v2Runtime) Execute(ctx context.Context, spec *jobv1.JobSpec) (jobs.Process, error) {
+	cmdSpec := spec.GetCommand()
+
+	// generate a uuid for the process, but encode it in the raw hex format.
+	// this makes it slightly easier to use from the command line, since many
+	// terminals treat '-' as a word separator, making it difficult to copy
+	// the whole string using a double-click.
+	u := uuid.New()
+	id := hex.EncodeToString(u[:])
+
+	cmd := exec.CommandContext(ctx, cmdSpec.GetCommand(), cmdSpec.GetArgs()...)
+	cmd.Env = append(os.Environ(), cmdSpec.GetEnv()...)
+
+	streamBuf := util.NewStreamBuffer()
+	done := make(chan struct{})
+
+	cmd.Stdout = streamBuf
+	cmd.Stderr = streamBuf
+	cmd.Stdin = nil
+	cmd.Cancel = func() error {
+		lg := slog.With("id", id)
+		lg.Debug("context canceled; attempting graceful shutdown")
+		streamBuf.Close() // NB: leaving this open will cause cmd.Wait to hang
+		start := time.Now()
+		go func() {
+			timeout := time.NewTimer(gracePeriod)
+			defer timeout.Stop()
+			select {
+			case <-timeout.C:
+				lg.Warn("process did did not exit within grace period, sending SIGKILL")
+				if err := cmd.Process.Signal(syscall.SIGKILL); err != nil {
+					lg.Error("failed to send SIGKILL", "error", err)
+				}
+			case <-done:
+				lg.Debug("process exited within grace period", "took", time.Since(start))
+			}
+		}()
+		return cmd.Process.Signal(syscall.SIGTERM)
+	}
+
+	job := &v2Process{
+		id:         id,
+		streamBuf:  streamBuf,
+		cmd:        cmd,
+		cmdContext: ctx,
+		done:       done,
+		status: &jobv1.JobStatus{
+			State:   jobv1.State_Pending,
+			Message: jobv1.State_Pending.String(),
+			Spec:    spec,
+		},
+	}
+
+	if err := l.configureCgroup(job, id, spec.GetLimits()); err != nil {
+		job.status.State = jobv1.State_Failed
+		job.status.Message = err.Error()
+		return nil, err
+	}
+
+	job.start()
+
+	return job, nil
+}
+
+// configureCgroup configures cgroup limits for the job.
+func (l *v2Runtime) configureCgroup(job *v2Process, id string, limits *jobv1.ResourceLimits) error {
+	path, err := l.mgr.CreateCgroupWithLimits(id, limits)
+	if err != nil {
+		return fmt.Errorf("failed to create cgroup for job %s: %w", id, err)
+	}
+	cf, err := os.OpenFile(path, syscall.O_RDONLY, 0)
+	if err != nil {
+		switch {
+		case errors.Is(err, syscall.EBUSY):
+			return fmt.Errorf("%w: cgroup %s has domain controllers enabled in cgroup.subtree_control", err, path)
```

Comment:
Did you mean "threaded controllers" here?

Reply:
From what I understand, "domain controllers" is correct in this context. From `man clone3`:
```
EBUSY (clone3() only)
    CLONE_INTO_CGROUP was specified in cl_args.flags, but the file descriptor specified in cl_args.cgroup refers to a version 2 cgroup in which a domain controller is enabled.

```

Reply:
Isn't that a possible error from the clone - i.e. from `cmd.Start`? Can you get the same error when opening the directory to prepare the file descriptor?

Reply:
Oh, oops - this must have been left from a refactor or something. you're right, those errors make no sense in this context!